### PR TITLE
Refactor sort checking to deal with synthesis for constructors

### LINKED ISSUE #895: Better sort checking for constructors
Right now, the following fails:

```rust
use flux_rs::*;

#[flux::refined_by(start: T, end: T)]
struct Range<T> {
    #[flux::field(T[start])]
    pub start: T,
    #[flux::field(T[end])]
    pub end: T,
}

#[flux::sig(fn(r: Range<i32>{v: Range { start: 0, end: 1 } == v }))]
fn foo2(_r: Range<i32>) {}
```

Result: 

```rust
  --> ../example.rs:11:33
   |
11 | #[flux::sig(fn(r: Range<i32>{v: Range { start: 0, end: 1 } == v }))]
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `?1s`, found constructor
 ```
 
 A nice solution for this would be (from @nilehmann): 
 
 The nice solution for this problem is whenever we check `Range { ... }` against a sort `s` we unify it against `Range<?1s>` (for a fresh ?1s). And then check fields assuming `?1s` . If `s` is `Range<int>` we will learn `Range<?1s> = Range<int> ==> ?1s = int` and consequently check fields against int. If `s` is `?0s` we only learn `?0s = Range<?1s>` and will check fields against `?1s`.

### PR DESCRIPTION
**Pull request by vrindisbacher**
closes #895

### CODE REVIEW

File: crates/flux-fhir-analysis/src/wf/sortck.rs [hunk 1]
```diff
@@ -270,8 +263,58 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
...
+                let mut used_fields = FxHashSet::default();
+                for expr in field_exprs.iter() {
+                    // check each expression against the sort
+                    // which will unify inferred variables
+                    if let Some(sort) = sort_by_name.get(&expr.ident.name) {
+                        self.check_expr(&expr.expr, sort)?;
+                    } else {
+                        return Err(self.emit_field_not_found(&sort, expr.ident));
+                    }
+                    if let Some(old_field) = used_fields.replace(expr.ident) {
```
Comment (vrindisbacher) [inline crates/flux-fhir-analysis/src/wf/sortck.rs:293]: @nilehmann - I'm kinda annoyed that we have to do this in both checking and synthesis. It feels like this shouldn't really be here but I'm not sure there's a way around it.

File: crates/flux-fhir-analysis/src/wf/sortck.rs [hunk 2]
```diff
@@ -270,8 +263,58 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
...
+                        return Err(
+                            self.emit_err(errors::DuplicateFieldUsed::new(expr.ident, old_field))
+                        );
+                    }
+                }
+                // check the spread against the sort
+                if let Some(spread) = spread {
+                    self.check_expr(&spread.expr, &sort)?;
+                } else if sort_by_name.len() != used_fields.len() {
```
Comment (vrindisbacher) [inline crates/flux-fhir-analysis/src/wf/sortck.rs:302]: Same with this

File: crates/flux-fhir-analysis/src/wf/sortck.rs [hunk 3]
```diff
@@ -270,8 +256,58 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
...
+                // first get the sort based on the path - for example S { ... } => S
+                // and we should expect sort to be a struct or enum app
+                let path_def_id = match path.res {
+                    ExprRes::Struct(def_id) | ExprRes::Enum(def_id) => def_id,
+                    _ => span_bug!(expr.span, "unexpected path in constructor"),
+                };
+                let sort_def = match self.genv.adt_sort_def_of(path_def_id) {
+                    Ok(sd) => sd,
+                    Err(_) => span_bug!(path.span, "Unexpected path in constructor"),
+                };
```
Comment (nilehmann) [inline crates/flux-fhir-analysis/src/wf/sortck.rs:270]: This error needs to be propagated

```suggestion
                let sort_def = match self.genv.adt_sort_def_of(path_def_id).emit(&self.genv)?;
```

File: crates/flux-fhir-analysis/src/wf/sortck.rs (hunk 1 above)
Comment (nilehmann) [reply crates/flux-fhir-analysis/src/wf/sortck.rs:286]: yes, I agree. We could move part of the logic to another function that only checks if all fields are present.

```rust
fn validate_constructor_fields(&self, sort_def: &AdtSortDef, exprs: &[fhir::FieldExpr], has_spread: bool);
```

File: crates/flux-fhir-analysis/src/wf/sortck.rs [hunk 4]
```diff
@@ -270,8 +279,42 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
...
+                    // check each expression against the sort
+                    // which will unify inferred variables
+                    if let Some(sort) = sort_by_name.get(&expr.ident.name) {
+                        self.check_expr(&expr.expr, sort)?;
+                    }
+                }
+                // check the spread against the sort
+                if let Some(spread) = spread {
+                    self.check_expr(&spread.expr, &sort)?;
+                }
```
Comment (nilehmann) [inline crates/flux-fhir-analysis/src/wf/sortck.rs:314]: This is the same as `check_field_expr`, no? Maybe we don't need `check_constructor_shape` and could call the old `check_field_expr`.

Comment (vrindisbacher) [reply crates/flux-fhir-analysis/src/wf/sortck.rs]: Yea it is - good point.

File: crates/flux-fhir-analysis/src/wf/sortck.rs [hunk 5]
```diff
@@ -270,8 +256,36 @@ impl<'genv, 'tcx> InferCtxt<'genv, 'tcx> {
...
+                // first get the sort based on the path - for example S { ... } => S
+                // and we should expect sort to be a struct or enum app
+                let path_def_id = match path.res {
+                    ExprRes::Struct(def_id) | ExprRes::Enum(def_id) => def_id,
+                    _ => span_bug!(expr.span, "unexpected path in constructor"),
+                };
+                let sort_def = match self.genv.adt_sort_def_of(path_def_id) {
+                    Ok(sd) => sd,
+                    Err(_) => span_bug!(path.span, "Unexpected path in constructor"),
+                };
```
Comment (nilehmann) [inline crates/flux-fhir-analysis/src/wf/sortck.rs:270]: I think this got reverted. The error should be propagated here.

**Review by nilehmann — APPROVED** [review]

**Pull request merged**